js/tji-datasets-explore.js

```diff
@@ -390,17 +447,64 @@ TJIChartView.prototype.filter_data = function() {
     return true;
   })
 
-  this.update_charts(data);
+  this.state.filtered_data = data;
 }
 
-TJIChartView.prototype.update_charts = function(data) {
-  this.update_record_count(data);
+TJIChartView.prototype.update_charts = function() {
+  var that = this;
   _.each(this.state.charts, function(chart){
-    chart.update(data);
+    chart.update(that.state.filtered_data);
   })
 }
 
-TJIChartView.prototype.update_record_count = function(data) {
+TJIChartView.prototype.update_record_count = function() {
   if (this.state.$record_count) jQuery(this.state.$record_count).remove();
-  this.state.$record_count = jQuery(this.record_count_template.replace('{count}', data.length)).prependTo(this.charts_elt_id);
+  this.state.$record_count = jQuery(this.record_count_template.replace('{count}', this.state.filtered_data.length)).prependTo(this.state.$above_charts_area);
+}
+
+TJIChartView.prototype.download = function() {
+  // Download all the data the user is currently viewing (that is,
+  // with their current filters applied).
+
+  // Note: This gets tricky because we want to download
+  // not only the fields available in the charts, but ALL
+  // fields (columns) in the complete dataset.
+
+  // Make note of the ids of our current (filtered) records.
```

Comment:
same as note above, this logic is troublesome for me to follow easily. How do you feel about keeping a list of filtered_ids in the state and then using those as a look up for the full data set object that's comprised of unique key(row key)-value(row data) pairs. We could remove the reference to filtered_data in state.